Refactored Data to be more robust.  Updated UI to be compatible.

JohnLeeroy opened the pull request.

ToxicBakery commented on app/src/main/java/com/codefororlando/petadoption/data/util/AnimalJsonDeserializer.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+import java.util.UUID;
+
+public class AnimalJsonDeserializer implements JsonDeserializer<Animal> {
```
Why bother with the custom deserializer?

ToxicBakery commented on app/src/main/java/com/codefororlando/petadoption/data/util/RetrievableImplJsonDeserializer.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.lang.reflect.Type;
+
+public class RetrievableImplJsonDeserializer implements JsonDeserializer<RetrievableImpl> {
```
Same as before, just asking for bugs

ToxicBakery commented:
So the more appropriate explanation against the deserializers is that they are specific your custom JSON model which will almost undoubtedly end up not matching the API. Even if they do end up matching, the custom deserializers are really only necessary when normal deserialization fails, usually for a complex reason, or because you have a *lot* of JSON to serialize or deserialize. Even then it would likely be best to move away from JSON to something faster vs writing deserializers as they are just extra work.

JohnLeeroy commented:
@ToxicBakery 
Thanks for the review, I removed the deserializers and modified the json file to deserialize nicely.  I just had to add a custom deserializer for Uri.  

I agree that custom serializer/deserializer is undesirable in most cases.  I dislike the idea of maintaining them.  It seems I was lazy with how I structured my json.

ToxicBakery commented:
Yeah, Uri is a good case for when you have to use one.
